design/one-pager-k8s-native-providers.md

```diff
@@ -0,0 +1,562 @@
+metadata:
+  name: demo-k8s-provider
+  namespace: my-infra-dev
+spec:
+  credentialsSecretRef:
+    name: demo-cluster-creds
+```
+
+2. Create a non-portable Rook CockroachDB `ClusterClass` that references the
+   `Provider` in its `providerRef`:
+
+```yaml
+apiVersion: cockroachdb.rook.crossplane.io/v1alpha1
+kind: ClusterClass
+metadata:
+  name: rook-cockroachdb
+  namespace: rook-infra-dev
+specTemplate:
+  providerRef:
+    name: demo-k8s-provider
+    namespace: my-infra-dev
+  # full documentation on all available settings can be found at:
+  # https://rook.io/docs/rook/master/cockroachdb-cluster-crd.html
+  scope:
+    nodeCount: 3
+    # You can only have one PersistentVolumeClaim in this list!
+    volumeClaimTemplates:
+    - metadata:
+        name: rook-cockroachdb-data
+      spec:
+        accessModes: [ "ReadWriteOnce" ]
+        # Uncomment and specify your StorageClass, otherwise
+        # the cluster admin defined default StorageClass will be used.
+        #storageClassName: "your-cluster-storageclass"
+        resources:
+          requests:
+            storage: "1Gi"
+  network:
+    ports:
+    - name: http
+      port: 8080
+    - name: grpc
+      port: 26257
+  secure: false
+  cachePercent: 25
+  maxSQLMemoryPercent: 25
+  # A key/value list of annotations
+  annotations:
+  #  key: value
+```
+
+3. Create a portable `PostgreSQLInstanceClass`:
+
+```yaml
+apiVersion: database.crossplane.io/v1alpha1
+kind: PostgreSQLInstanceClass
+metadata:
+  name: postgresql-standard
+  namespace: app-project1-dev
+classRef:
+  kind: ClusterClass
+  apiVersion: cockroachdb.rook.crossplane.io/v1alpha1
+  name: rook-cockroachdb
+  namespace: rook-infra-dev
+```
+
+4. Create a `PostgreSQLInstance` claim:
+
+```yaml
+apiVersion: database.crossplane.io/v1alpha1
+kind: PostgreSQLInstance
+metadata:
+  name: app-postgresql
+  namespace: app-project1-dev
+spec:
+  classRef:
+    name: postgresql-standard
+  writeConnectionSecretToRef:
+    name: postgresqlconn
+  engineVersion: "9.6"
+```
+
+On creation of the `PostgreSQLInstance` claim, the claim controller would
+configure and create a `Cluster` (`cockroachdb.rook.crossplane.io/v1alpha1`)
+instance. Then, the managed controller would use the Crossplane `Cluster`
+instance to create a `Cluster` (`cockroachdb.rook.io/v1alpha1`) in the target
+Kubernetes cluster where the Rook CockroachDB operator is installed. This
+workflow is similar to dynamic provisioning in a typical infrastructure
+provider:
+
+![K8s Native Providers](./images/k8s-native-providers.png)
+
+### Non-Portable Resources
+
+Some resources exposed by Kubernetes-native infrastructure providers may not be
+initially supported by dynamic provisioning. This is true for existing
+infrastructure stacks as well, as types like `VirtualNetwork` in Azure can be
+provisioned statically, but not in the dynamic `claim <-> class` pattern. These
+types of resources may eventually be supported by portable claim types in core
+Crossplane.
```

Comment:
I think there's something more subtle going on here than support for dynamic provisioning. Specifically:

* You can't claim these managed resources.
* You can't dynamically provision these managed resources, because dynamic provisioning is triggered by a claim.

This raises the question "what is the purpose of a claim?" Is it just a request to dynamically provision a managed resource, or is it a request to "claim" a managed resource in the sense of claiming an exclusive right to its use, indicating that it's no longer available to satisfy other claims.

In the Kubernetes PV/PVC pattern we're modeled on you can't (easily) use a PersistentVolume without claiming it. So whether a volume is provisioned dynamically or statically, you must always claim it. I think we must follow this pattern for resource kinds that _can_ be claimed (e.g. `CloudSQLInstance`). If we were to start using them without ever claiming them there would be no way for a `MySQLInstance` claim to distinguish whether an unclaimed `CloudSQLInstance` was or was not actually in use. The lack of a claim binding is taken to be a signal that the `CloudSQLInstance` is available for use.

We're kind of punting on what to do about these managed resource kinds that can't be claimed for the time being. I'm not sure there's an actual action to be taken in the context of this design doc, but I wanted to communicate that I think of these managed resources as more in some kind of "unclaimable danger zone" rather than simply "not dynamically provisionable".

Reply:
I definitely understand the concern here, but one thing to keep in mind is that most of these resources (e.g. virtual networks, subnets, object stores, etc.) are resources that could be "used" by many different apps / services / etc. For instance, I may want to use Crossplane to manage multiple MySQL database instances, each for a different application, that all exist in the same network. If I can't provision resources in a network unless I can claim an unbound `VirtualNetwork`, then I can only use that `VirtualNetwork` for a single application. This differs from something like a `CloudSQLInstance` which is likely only used by a single application. One way we could potentially allow for claiming these currently "unclaimable" resources would be to allow multiple claims to bind to them. However, this would also be divorced from the PV/PVC pattern, so I am not sure this is an improvement.

Reply:
Perhaps "exclusive right to use" was not the right wording for what a claim represents. In the case of persistent volume claims it's more a sign that "it is being used". There is a strict 1:1 PV:PVC relationship, but I don't think there is a strict 1:1 PVC:Pod relationship.

I'm not really making the case that we need something like a `NetworkClaim` so much as pointing out that we should probably form a stance on these "unclaimable" resources before they proliferate further, and that agreeing on what exactly the point of a claim is (is it simply a request for dynamic provisioning?) could help do that.

Reply:
Just FYI, the way that PVCs express their intended consumption is with an `accessModes` field on their spec, e.g.:
```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: mypvc
spec:
  accessModes:
  - ReadWriteOnce
```

If multiple consumers are allowed/expected, then the accessModes value would be `ReadWriteMany`.